core/src/test/scala/org/apache/spark/rdd/AsyncRDDActionsSuite.scala

```diff
@@ -197,4 +197,30 @@ class AsyncRDDActionsSuite extends SparkFunSuite with BeforeAndAfterAll with Tim
       Await.result(f, Duration(20, "milliseconds"))
     }
   }
+
+  test("SimpleFutureAction callback must not consume a thread while waiting") {
+    val executorInvoked = Promise[Unit]
+    val fakeExecutionContext = new ExecutionContext {
+      override def execute(runnable: Runnable): Unit = {
+        executorInvoked.success(())
+      }
+      override def reportFailure(t: Throwable): Unit = ???
+    }
+    val f = sc.parallelize(1 to 100, 4).mapPartitions(itr => {Thread.sleep(1000L); itr}).countAsync()
+    f.onComplete(_ => ())(fakeExecutionContext)
+    assert(!executorInvoked.isCompleted)
```

Comment:
if I understand right, the idea of this is test is you set up some really slow transformations, so that it won't actually complete before you get to this assert?  at the very least lets add a comment saying that.  I worry that this will become a flaky test ... though I can't immediately think of an alternative.
Also -- what happens to the async actions at the end of this?  Does it get cleaned up nicely?